docs/design/sql/05-dml.md

```diff
@@ -0,0 +1,100 @@
+- **CREATE** which creates new record(s) (`INSERT`)
+- **UPDATE** which updates record(s) fields (`UPDATE`, `MERGE`)
+- **DELETE** which deletes record(s) (`DELETE`)
+
+## 1. INSERT
+To create new record(s) one of two variations of `INSERT` statement can be used:
+- `INSERT INTO` - will append to the target table. If it’s not possible due to unique constraint violation, it will
+  fail. It’s a safe choice if the target never overwrites (such as Kafka, JMS or a JDBC that use a sequence for the
+  primary key or has no primary key)
+- `INSERT OVERWRITE` - will allow overwrites. It will behave as a sequence of `DELETE` followed by an `INSERT`,
+  performed atomically. Specifically, the fields not listed in the insert statement will not retain their values, but
+  will be assigned `null` or default values. If `INSERT OVERWRITE` is used for a target that doesn’t overwrite by
+  design (like Kafka), it will behave exactly as `INSERT INTO`. No error will be thrown. The delete in the imaginary
+  `DELETE` action will delete no rows.
+
+### 1.1 Serialization
+In addition to primitive SQL types (which can be mapped to Java types and stored directly), two serialization formats
+are supported:
+- POJO (reflective access to public properties and fields) - requires Java class on the classpath and no-arg
+  constructor
+- Portable (supported just for IMDG data structures) - if `InMemoryFormat` is set to `BINARY` it requires just
+  `ClassDefinition` to be registered upfront, if `InMemoryFormat` is set to `OBJECT` it requires also Portable factory
+  and Portable class itself on the classpath
+
+Serialization formats have following precedence (the order in which we check whether given serialization can be
+applied):
+1. primitive SQL types
+2. POJO
+3. Portable
+
+Eventual errors from type mismatch between declared and actual types are deferred to statement execution.
+
+The way the data is serialized is expressed via DDL. Depending on the nature of stored objects that information can be
+encoded either in column names or using `OPTIONS` clause.
+
+#### 1.1.1 Key-Value storage
+The serialization formats of key and value are specified separately. To determine whether given field belongs to key or
+value either Java class or Portable `ClassDefinition` is inspected. In case of a name clash between fields of key and
+value composite objects, the ones from key take precedence and are actually written.
+
+To define a table where both key and value are of primitive SQL types, `__key` & `this` as column names should be used.
+For example:
+```
+CREATE EXTERNAL TABLE name (
+  __key INT,
+  this VARCHAR
+) TYPE "com.hazelcast.LocalPartitionedMap"
+```
+
+Given:
+```
+class PersonId {
+    public int id;
+}
+
+class Person {
+    public String name;
+}
+```
+To define a table where both key and value are POJOs and respectively of `PersonId` & `Person` class, `keyClass` &
+`valueClass` should be used. For example:
+```
+CREATE EXTERNAL TABLE name (
```

Comment:
Here we basically duplicate the definition of columns. First, they are defined explicitly. Second, they could be derived from the class. The idea with `CREATE MAP / ALTER MAP` addresses this - you just provide class definitions (in the form of a class name, or portable class IDs), and columns are extracted from there. 

Is there any important use cases for us where explicit column definition might be needed? E.g. to define an alias for the column, or to mark it non-nullable, etc?

Reply:
There is one important use case where the map does not exist and user would like to start inserting records to it. In any other scenario, in theory, we don't need any `CREATE`/`ALTER` as we can derive the schema from existing entries. We could for instance allow `CREATE MAP` (however it's another customization to the syntax) only if map with given name does not exist - failing otherwise. Moreover, if we want to avoid duplication of defined columns we would need to take that customization a little bit further and disallow specifying table columns at all, something like `CREATE MAP map_name OPTIONS (keyClass 'PersonId', valueClass 'Person')`.

The second use case I see, where specifying (and not deriving from existing entries) table columns might be useful is usage of polymorphic maps. Maybe, user would like to be specific (and sure) what classes are read and written - however I'm not sure how widespread such usage is.